src/ray/util/size_literals.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// These user-defined literals makes sizes.
+
+#pragma once
+
+#include <cstdint>
+
+constexpr size_t operator""_B(size_t sz) { return sz; }
+
+constexpr size_t operator""_KiB(size_t sz) { return sz * 1024ULL; }
+constexpr size_t operator""_KB(size_t sz) { return sz * 1000ULL; }
+
+constexpr size_t operator""_MiB(size_t sz) { return sz * 1024_KiB; }
+constexpr size_t operator""_MB(size_t sz) { return sz * 1000_KB; }
+
+constexpr size_t operator""_GiB(size_t sz) { return sz * 1024_MiB; }
+constexpr size_t operator""_GB(size_t sz) { return sz * 1000_MB; }
+
+constexpr size_t operator""_TiB(size_t sz) { return sz * 1024_GiB; }
+constexpr size_t operator""_TB(size_t sz) { return sz * 1000_GB; }
+
+constexpr size_t operator""_PiB(size_t sz) { return sz * 1024_TiB; }
+constexpr size_t operator""_PB(size_t sz) { return sz * 1000_TB; }
+
+constexpr size_t operator""_KiB(long double sz) {
+  const long double res = sz * 1024ULL;
+  return static_cast<size_t>(res);
+}
+constexpr size_t operator""_KB(long double sz) {
+  const long double res = sz * 1000ULL;
+  return static_cast<size_t>(res);
+}
+
+constexpr size_t operator""_MiB(long double sz) {
+  const long double res = sz * 1024_KiB;
+  return static_cast<size_t>(res);
+}
+constexpr size_t operator""_MB(long double sz) {
+  const long double res = sz * 1000_KB;
+  return static_cast<size_t>(res);
+}
+
+constexpr size_t operator""_GiB(long double sz) {
+  const long double res = sz * 1024_MiB;
+  return static_cast<size_t>(res);
+}
+constexpr size_t operator""_GB(long double sz) {
+  const long double res = sz * 1000_MB;
+  return static_cast<size_t>(res);
+}
+
+constexpr size_t operator""_TiB(long double sz) {
+  const long double res = sz * 1024_GiB;
+  return static_cast<size_t>(res);
+}
+constexpr size_t operator""_TB(long double sz) {
+  const long double res = sz * 1000_GB;
+  return static_cast<size_t>(res);
+}
+
+constexpr size_t operator""_PiB(long double sz) {
+  const long double res = sz * 1024_TiB;
```

Comment:
[nit] why not:

constexpr size_t operator""_PiB(long double sz) {
  const long double res = sz * 1_PiB;
  return static_cast<size_t>(res);
}

Reply:
and these double-to-size_t code are so duplicated, I wonder if we can make it a macro.

```
// Macro to generate (long double -> size_t) literal from (size_t -> size_t)
#define CREATE_FLOAT_SIZE_LITERAL(suffix) \
  constexpr size_t operator""_##suffix(long double sz) { \
    return static_cast<size_t>(sz * 1_##suffix); \
  }

// Using the macro to define long double -> size_t literals
CREATE_FLOAT_SIZE_LITERAL(KiB)
CREATE_FLOAT_SIZE_LITERAL(MiB)
CREATE_FLOAT_SIZE_LITERAL(GiB)
CREATE_FLOAT_SIZE_LITERAL(TiB)
CREATE_FLOAT_SIZE_LITERAL(PiB)
```

Reply:
> [nit] why not:
constexpr size_t operator""_PiB(long double sz) {
const long double res = sz * 1_PiB;
return static_cast<size_t>(res);
}

Updated.

Reply:
I'm not a fan of macros.

Using macro is generally not recommended (effective C++ item-1 has a few good arguments):
- Hard to debug;
- Affective, especially in header file.

I almost only use macros at error propagation (i.e. `ray/util/logging.h`) to save some boilerplate code for unhappy path.